Net 2714/xw cli read command

wangxinyi7 opened the pull request:
### Description

Implement the resource read command to allow cli to talk with resource back end.

[Here](https://whimsical.com/resource-read-flag-validation-LGRopyidXTqQzGsiUwy1NZ) is the flow to validate the flags

**Usage:**

You could provide resource information in the command line directly or wrap them up in the hcl file, but user can not use them together:

1. consul resource read [type] [name] -partition=<default> -namespace=<default> -peer=<local>
2. consul resource read -f test.hcl

**Test coverage:**

This PR only addressed arguments validation.
After the apply command gets ready, it will coverage some functionality test.

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

JadhavPoonam commented on internal/resource/http/http.go:
```diff
@@ -176,10 +176,14 @@ func (h *resourceHandler) handleDelete(w http.ResponseWriter, r *http.Request, c
 		Partition: query.Get("partition"),
-		PeerName:  query.Get("peer_name"),
-		Namespace: query.Get("namespace"),
+		PeerName:  query.Get("peer"),
```
I thought we would be supporting both `peer` and `peer_name` 🤔

wangxinyi7 replied:
that's right! just corrected it

JadhavPoonam commented on command/resource/resource.go:
```diff
@@ -0,0 +1,47 @@
+
+Read a resource:
+
+$ consul resource read [type] [name] -partition=<default> -namespace=<default> -peer=<local> -consistent=<false> -json
```
Did we miss instantiating the `consistent` flag?

JadhavPoonam replied:
It is best to stick with `stale` flag if other CLI commands are using `stale` instead of `consistent`. can you check?

wangxinyi7 replied:
right, `stale` is a http flag already, I think we need to stick with it instead of using a new name

https://github.com/hashicorp/consul/blob/4f9955d91e1f7cdab68ab407d4789e1339ff6bfa/command/flags/http.go#L73

wangxinyi7 replied:
Just updated the code

dhiaayachi commented on command/flags/http.go:
```diff
@@ -174,3 +179,9 @@ func (f *HTTPFlags) AddPartitionFlag(fs *flag.FlagSet) {
 			"Admin Partitions are a Consul Enterprise feature.")
 }
+
+func (f *HTTPFlags) AddPeerName() *flag.FlagSet {
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	fs.Var(&f.peer, "peer", "Specifies the name of peer to query. By default, it is `local`.")
+	return fs
+}
```
Why is this added specifically for the resource API? I'm wondering if we should hold on for peering as it's not clear how we will implement the api? 🤔

wangxinyi7 replied:
@dhiaayachi 
I'm not sure how the `peer` flag is used by the backend, I added this flag here based on the AC of the ticket, it basically requires the `peer` information, that's why I added it here.

Please let me know if the information here is redundant or not.

JadhavPoonam replied:
@dhiaayachi The `peer` field is part of the resource gRPC API hence it's included in both the HTTP side and the CLI side. 
The pending discussion/work is about whether peer should be moved outside the tenancy block but I am not sure what you mean by "we are not clear how we will implement the api" can you elaborate?

dhiaayachi replied:
I agree I think supporting both make sense.

dhiaayachi commented on command/resource/read/read.go:
```diff
@@ -0,0 +1,176 @@
+	c.help = flags.Usage(help, c.flags)
+}
+
+func (c *cmd) Run(args []string) int {
```
It would be useful to have examples of the command usage in the PR, I'm not sure I understand all of the flags logic in here

wangxinyi7 replied:
I updated the usage information [here](https://github.com/hashicorp/consul/blob/47bdf63c7e773e9d065014b8d8a0839351d0b3ee/command/resource/read/read.go#L172-L196)

You could trigger the help message by executing `consul resource read -h`

dhiaayachi commented on command/resource/read/read.go:
```diff
@@ -0,0 +1,176 @@
+	var resourceName string
+	var opts *api.QueryOptions
+
+	if args[0] == "-f" {
```
Should we check of len(args)>0 in here to avoid a panic?

wangxinyi7 replied:
Good catch! I'll add the sanity check here

dhiaayachi commented on command/resource/read/read.go:
```diff
@@ -0,0 +1,176 @@
+			return 1
+		}
+
+		inputArgs := args[2:]
```
Same here about len check

wangxinyi7 replied:
I just tested and it's gonna be at least empty, and the response is `404` since it doesn't provide specific tenancy information

dhiaayachi replied:
I think it's better if that sanity check is extracted at the higher level to have better readability, wyt?

dhiaayachi commented on command/resource/read/read.go:
```diff
@@ -0,0 +1,176 @@
+
+	if args[0] == "-f" {
+		if err := c.flags.Parse(args); err != nil {
+			if err == flag.ErrHelp {
```
I know this is copied from elsewhere, but a better check here would be `errors.Is(err, flag.ErrHelp)` to avoid missing wrapped errors, in case newer versions of the library use it.

wangxinyi7 replied:
Thanks @dhiaayachi
Today I learned! Useful and important!
👍

dhiaayachi commented on command/resource/read/read_test.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	cases := map[string]tc{
+		"missing file path": {
```
I think we are missing empty and nil args test cases in here.

wangxinyi7 replied:
right! just as you pointed out the length sanity check! will add it here

dhiaayachi replied:
What about `nil` args?

dhiaayachi commented on internal/resource/http/http.go:
```diff
@@ -176,10 +176,18 @@ func (h *resourceHandler) handleDelete(w http.ResponseWriter, r *http.Request, c
+	if namespace == "" {
+		namespace = query.Get("ns")
+	}
+	peer := query.Get("peer")
+	if peer == "" {
+		peer = query.Get("peer_name")
+	}
```
Have we agreed to changing `peer_name` to `peer`?

JadhavPoonam replied:
The HTTP API expects the query param to be `peer_name` which is consistent with the gRPC but the CLI [calls it](https://github.com/hashicorp/consul/blob/main/api/api.go#L125) `peer`. So we either support both or update CLI to add these [here](https://github.com/hashicorp/consul/blob/main/api/api.go#L834).

I am inclined to support both rather over an inconsistent experience within CLI. WDYT?

dhiaayachi replied:
I agree on supporting both to keep retro compatibility but also provide a way of having a consistent UX. Maybe we should also deprecate peer in preparation of removing it? 🤔

dhiaayachi reviewed (changes requested):
Thank you for working on this @wangxinyi7 and sorry for the late review 😞 
I had few comments inline.

JadhavPoonam reviewed (changes requested):
@wangxinyi7 I think the flags parsing needs some cleaning up.
I noticed when we pass the params in the `hcl` file the command throws an error since it expects resource type and name which really should be parsed from the file provided.
Example:
`consul resource read -f demo.hcl` returns `Your argument format is incorrect: Must specify two arguments: resource type and resource name`

Also, on a general note can you add some more details in the PR description with things like what changes this PR is proposing such as usage and how it was tested, etc?

JadhavPoonam commented on command/resource/read/read.go:
```diff
@@ -0,0 +1,176 @@
+			Partition:         c.http.Partition(),
+			Peer:              c.http.PeerName(),
+			Token:             c.http.Token(),
+			RequireConsistent: c.http.Stale(),
```
Wouldn't stale be the opposite of consistent? 🤔

wangxinyi7 replied:
Thank you!

wangxinyi7 commented:
> @wangxinyi7 I think the flags parsing needs some cleaning up. I noticed when we pass the params in the `hcl` file the command throws an error since it expects resource type and name which really should be parsed from the file provided. Example: `consul resource read -f demo.hcl` returns `Your argument format is incorrect: Must specify two arguments: resource type and resource name`
> 
> Also, on a general note can you add some more details in the PR description with things like what changes this PR is proposing such as usage and how it was tested, etc?

Thanks for your comment. @JadhavPoonam 

- I updated the validation logic behind the scene, I'm also aware of the logic here is messy, please leave comment to see if we could simplify it
- I ran the test locally and found out it works for me, we could sync offline on this in case we overlooked something
- updated the PR description with more info in it.

JadhavPoonam commented:
> > @wangxinyi7 I think the flags parsing needs some cleaning up. I noticed when we pass the params in the `hcl` file the command throws an error since it expects resource type and name which really should be parsed from the file provided. Example: `consul resource read -f demo.hcl` returns `Your argument format is incorrect: Must specify two arguments: resource type and resource name`
> > Also, on a general note can you add some more details in the PR description with things like what changes this PR is proposing such as usage and how it was tested, etc?
> 
> Thanks for your comment. @JadhavPoonam
> 
> * I updated the validation logic behind the scene, I'm also aware of the logic here is messy, please leave comment to see if we could simplify it
> * I ran the test locally and found out it works for me, we could sync offline on this in case we overlooked something
> * updated the PR description with more info in it.

Thanks @wangxinyi7 ! I ran this again and the issue is gone perhaps I had a stale branch, sorry about that.

dhiaayachi reviewed (commented):
@wangxinyi7 I went through the changes, thank you for considering my comments. I have one last comment open, about a test case for nil args.

My second comment is a Nit so not blocking.

wangxinyi7 commented:
> @wangxinyi7 I went through the changes, thank you for considering my comments. I have one last comment open, about a test case for nil args.
> 
> My second comment is a Nit so not blocking.

@dhiaayachi Thanks for doing the second round of review! Not sure why, but the reply text box disappeared from my side, so I'll address your three comments here:

1. Yes, I lifted up the sanity check outside of the function call for readability, thanks!
2. Yes, I added the nil check, thanks!
3. I think the obstacle here is that the CLI side `api.QueryOptions` hardcoded the `peer` while the grpc uses `peer_name` AND they've been already extensively adopted by both sides. So I'm afraid for now we have to be compatible to them to make them happy 😅

wangxinyi7 commented on command/resource/read/read.go:
```diff
@@ -0,0 +1,198 @@
+				Namespace: parsedResource.Id.Tenancy.GetNamespace(),
+				Partition: parsedResource.Id.Tenancy.GetPartition(),
+				Peer:      parsedResource.Id.Tenancy.GetPeerName(),
+				Token:     c.http.Token(),
```
Good catch! updated

dhiaayachi reviewed (approved):
Thank you for integrating my comments @wangxinyi7!

🚀 🚀 🚀
